File: video/api/worker.py
"""In-process worker pool (doc 02 scale dial: minutes tier).

A small thread pool drains a queue of jobs, runs the matching stage runner, and writes
the result/error back to the durable store. Two workers by default so voiceover/transcript
can overlap the long image job (doc 09 'pipeline overlap'); image gen itself is GPU-serial.

To scale to the hours tier (doc 06), swap this for RQ/arq + Redis — the store and the
runners are unchanged; only the queue/process boundary moves.
"""
from __future__ import annotations

import os
import queue
import threading
import traceback
from typing import Any, Callable

from .control import Cancelled, cancellation
from .jobs import JobStore

Runner = Callable[[JobStore, str, Any], dict]


class Worker:
    def __init__(self, store: JobStore, concurrency: int | None = None):
        self._store = store
        self._q: "queue.Queue[tuple[str, Runner, Any]]" = queue.Queue()
        self._threads: list[threading.Thread] = []
        self._n = concurrency or int(os.environ.get("VIDEO_WORKERS", "2"))

    def start(self) -> None:
        for i in range(self._n):
            t = threading.Thread(target=self._loop, name=f"video-worker-{i}", daemon=True)
            t.start()
            self._threads.append(t)

    def submit(self, job_id: str, runner: Runner, req: Any) -> None:
        self._q.put((job_id, runner, req))

    def _loop(self) -> None:
        while True:
            job_id, runner, req = self._q.get()
            try:
                if cancellation.is_cancelled(job_id):  # stopped while still queued
                    self._store.set_canceled(job_id)
                    continue
                self._store.set_status(job_id, "running")
                result = runner(self._store, job_id, req)
                self._store.finish(job_id, result)
            except Cancelled:
                self._store.set_canceled(job_id)
            except Exception as exc:  # noqa: BLE001 — surface any failure to the job
                if cancellation.is_cancelled(job_id):
                    self._store.set_canceled(job_id)
                else:
                    detail = str(exc) or traceback.format_exc(limit=3)
                    self._store.fail(job_id, detail)
            finally:
                cancellation.clear(job_id)
                self._q.task_done()

File: web/src/lib/jobs/worker.ts
import "server-only";

import { JOB_WORKER_CONCURRENCY } from "@/lib/config";
import { cancellation, Cancelled } from "@/lib/jobs/cancellation";
import * as store from "@/lib/jobs/store";

// In-process worker pool, port of video/api/worker.py. Drains a queue of jobs with
// bounded concurrency, runs the matching stage runner, and writes the result/error back
// to the durable store. Runs inside the same persistent Node server as the rest of the
// app (this deploys as `next build --standalone` + `node server.js`, not serverless),
// so — like the Python worker — an in-process queue survives for the process lifetime.

export type Runner<P = Record<string, unknown>> = (jobId: string, params: P) => Promise<Record<string, unknown>>;

type QueueItem = { jobId: string; runner: Runner; params: unknown };

class Worker {
  private queue: QueueItem[] = [];
  private active = 0;

  constructor(private concurrency: number) {}

  submit<P>(jobId: string, runner: Runner<P>, params: P): void {
    this.queue.push({ jobId, runner: runner as Runner, params });
    this.pump();
  }

  private pump(): void {
    while (this.active < this.concurrency && this.queue.length > 0) {
      const item = this.queue.shift();
      if (!item) break;
      this.active++;
      void this.runOne(item).finally(() => {
        this.active--;
        this.pump();
      });
    }
  }

  private async runOne({ jobId, runner, params }: QueueItem): Promise<void> {
    try {
      if (cancellation.isCancelled(jobId)) {
        await store.setCanceled(jobId);
        return;
      }
      await store.setStatus(jobId, "running");
      const result = await runner(jobId, params as Record<string, unknown>);
      await store.finish(jobId, result);
    } catch (err) {
      if (err instanceof Cancelled || cancellation.isCancelled(jobId)) {
        await store.setCanceled(jobId);
      } else {
        const detail = err instanceof Error ? err.message : String(err);
        await store.fail(jobId, detail);
      }
    } finally {
      cancellation.clear(jobId);
    }
  }
}

// Singleton on globalThis so Next.js dev-mode module reloads don't spawn a second queue.
const g = globalThis as unknown as { __videoWorker?: Worker };
export const worker = g.__videoWorker ?? (g.__videoWorker = new Worker(JOB_WORKER_CONCURRENCY));
